### PR TITLE
cmd/compile/internal/syntax: implement feature 'immediate return' for expression statements

### PR DESCRIPTION
**Pull request by 0x9n0p**
### Immediate Return

Golang has simple error handling, but sometimes it can lead us to write useless 'if' statements.

For example, if you call a function that gives you an error, and you want to return it when it is nil, you have to write an if statement to check 'err != nil' and return the error, whereas, by this new feature, you can add a question mark '?' after the right parents to do that!

```go
// This line
callee()?

// will be converted to these lines.
if err := callee(); err != nil {
    return err
}
```

### CODE REVIEW

File: src/cmd/compile/internal/syntax/nodes.go [hunk 1]
```diff
@@ -235,11 +235,12 @@ type (
...
-	// Fun(ArgList[0], ArgList[1], ...)
+	// Fun(ArgList[0], ArgList[1], ...)?
 	CallExpr struct {
-		Fun     Expr
-		ArgList []Expr // nil means no arguments
-		HasDots bool   // last argument is followed by ...
+		Fun       Expr
+		ArgList   []Expr // nil means no arguments
+		HasDots   bool   // last argument is followed by ...
+		ImmReturn bool   // callee()?
```
Comment (kehiy) [inline src/cmd/compile/internal/syntax/nodes.go]:
```suggestion
        ImmReturn bool // ImmReturn means expr returns immediately in error case
```

File: src/cmd/compile/internal/syntax/walk.go [hunk 2]
```diff
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// This file implements syntax tree walking.
+// This file implements syntax tree walking and changing.
```
Comment (kehiy) [inline src/cmd/compile/internal/syntax/walk.go:5]: Maybe it's better to not define walk and change as a part of walk file(?).

They defined it to walk over AST, not to change it.

**Review by kehiy — COMMENTED** [review]
Overall these changes are OK. I didn't run the code but from a quick review:

There was only 2 issues.

Also, in the term of compiler design I'm not sure if handling this kind of stuff in parser level or generating code(?) is a best practice. do you have any idea?

File: src/cmd/compile/internal/syntax/walk.go (hunk 2 above)
Comment (0x9n0p) [reply src/cmd/compile/internal/syntax/walk.go:5]: Their implementations are almost the same, and I couldn't find a new concept to create a new file. I think we should have them together in one file for now.

**Comment by 0x9n0p** [comment]
You are right. I think the best way to add our features is to extend the Golang implementation instead of modifying it. So that we simply merge upstream changes without resolving conflicts.
After the parser level, many optimizations will begin. **Changing the AST** should be done right after the parser generates the AST and before the optimizations.

@kehiy
> Also, in the term of compiler design I'm not sure if handling this kind of stuff in parser level or generating code(?) is a best practice. do you have any idea?

File: src/cmd/compile/internal/syntax/walk.go (hunk 2 above)
Comment (kehiy) [reply src/cmd/compile/internal/syntax/walk.go:5]: Ok, so let's consider it as a TODO for the future.

**Pull request merged**